File: src/main/java/ar/com/flow/bankaccount/Statement.java
package ar.com.flow.bankaccount;

import ar.com.flow.bankaccount.balance.Balance;
import ar.com.flow.bankaccount.transaction.Receipt;

import java.util.Optional;

public interface Statement {
    int total();

    Optional<Receipt> first();

    Optional<Receipt> last();

    void add(Receipt receipt);

    boolean contains(Receipt receipt);

    boolean containsInOrder(Receipt... records);

    Optional<Balance> getCurrentBalance();

    Optional<Balance> getPreviousBalance();

    Optional<Balance> getInitialBalance();
}

File: src/main/kotlin/ar/com/flow/bankaccount/Statement.kt
package ar.com.flow.bankaccount

import ar.com.flow.bankaccount.balance.Balance
import ar.com.flow.bankaccount.transaction.receipt.Receipt
import java.util.*

interface Statement {
    val currency: String
    fun total(): Int
    fun first(): Optional<Receipt>
    fun last(): Optional<Receipt>
    fun add(receipt: Receipt)
    operator fun contains(receipt: Receipt): Boolean
    fun containsInOrder(vararg records: Receipt): Boolean
    fun getInitialBalance(): Balance
    fun getCurrentBalance(): Balance
    fun getPreviousBalance(): Balance
    fun sum(numberOfTransactions: Int): Balance
}